### DISCUSSION TITLE
openra架构学习讨论贴

### DISCUSSION THREAD

**Discussion by miketang84** [openra-rs/learn-openra#26] [General]
openra架构学习讨论贴。

**Comment by minghuaw** [comment]
# OpenRA Mods, Actors, Traits 概览

详细的笔记和资料可以参考2021年四月21日的例会讨论内容([slides](https://github.com/openra-rs/learn-openra/blob/main/docs/minghuaw/Presentation_Apr_21_2021.pdf), [录像](https://www.bilibili.com/video/BV1P64y1i7bK?share_source=copy_web))。这里主要是做一个摘要以方便后续的讨论。为了方便以下的讨论，Trait代表着OpenRA当中的广义的trait，Trait包含了一个info class和一个trait class。

## 参考资料

[1] OpenRA wiki modding guide: https://github.com/OpenRA/OpenRA/wiki/Modding-Guide
[2] OpenRA Book: https://www.openra.net/book/glossary.html
[3] Delft Students On Software Architecture - OpenRA: https://delftswa.github.io/chapters/openra/
[4] OpenRA Traits Documentation: https://docs.openra.net/en/latest/release/traits/

## 大致的对应关系：

| OpenRA | Bevy | References |
|----------|------|------------|
| `Mods`  | `Plugin` | [[1](https://github.com/OpenRA/OpenRA/wiki/Modding-Guide)] |
| `Actor`  | `Entity`  | [[2](https://www.openra.net/book/glossary.html)] |
| Trait info class | `Component` | [[1](https://github.com/OpenRA/OpenRA/wiki/Modding-Guide)] |
| Trait trait class | `System` | [[1](https://github.com/OpenRA/OpenRA/wiki/Modding-Guide)] |

## 对于用 Bevy ECS 实现带来的问题

OpenRA虽然自称 `Actor`-info class-trait class 三者对应着`Entity`-`Component`-`System`，但是从OpenRA实现来看这个对应关系是存在问题的，下面简单总结一下目前发现的问题。

1.  info class定义上与`Component`的不同
> There is one instance of the infoclass shared across all actors of the same type. Each actor gets its own instance of the trait class itself. [1]

从这个定义来看，info class并不能直接等同于 `Component`。假如说有一群`DOG`，他们不可能共同拥有同一个Component（比如说速度和位置）。同时info class也负责创造（`Create()`）一个对应的trait class，这个也是超出了`Component`定义的内容的。

2. 继承所带来的复杂性（可以参考[笔记](https://github.com/openra-rs/learn-openra/blob/main/docs/minghuaw/openra_mods_actors_traits/OpenRA%20Mods%20%26%20Traits.md)当中的一个UML图）
所有OpenRA的 Trait的info class 都继承`TraitInfo`，而最终用于组装一个`Actor`的Trait可能已经经过了一长串的继承关系，这就带来了复杂性。其中可能有一些父类的properties在子类当中并不需要，还有就是对于某一个具体的`Actor`来说，这种复杂性增加了分析它的`Component`和`System`难度。

3. `TraitInfo` 和 `Trait` 之间的边界不清楚
目前通过一些Bevy ECS的例子来看，`Component` 和 `System` 那样有比较清楚的界限，数据都是`Component`并且`System`并不拥有数据，但是在OpenRA的实现当中`TraitInfo` 和 `Trait`都拥有数据。所以从数据的角度来看，`TraitInfo` 和 `Trait`与Bevy中的`Component`和`System`还不能一一对应。

## Side Notes

除了 Trait 之外，OpenRA内还有一个叫`Activity`的东西，并且有一些 trait class 会产生一些`Activity`。目前有限的research的情况下，这个`Activity`本质也是`System`但是更像是需要两个或者多个`Actor`互动的`System`。

## 下一步

个人认为下一步需要做的是

1. 制作一个最小可玩的OpenRA mod，这里可以参考其官方的 Modding guide [1]。这个mod只需要达到能够启动，地图中仅包含一个单位，然后这个单位可以做到一两个指令就可以。
2. 用一个OpenRA的`Actor`来分析如何可以把它用ECS系统来重构。
